service/subprocessmanager/manager.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package subprocessmanager
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"math"
+	"math/rand"
+	"os"
+	"os/exec"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/kballard/go-shellquote"
+	"go.uber.org/zap"
+)
+
+const (
+	// delayMutiplier is the factor by which the delay scales
+	delayMultiplier float64 = 2.0
+	// initialDelay is the initial delay before a process is restarted
+	initialDelay time.Duration = 1 * time.Second
+)
+
+// GetDelay will compute the delay for a given process according to its crash count and time alive using an exponential backoff algorithm
+func GetDelay(elapsed time.Duration, healthyProcessDuration time.Duration, crashCount int, healthyCrashCount int) time.Duration {
```

Comment:
Does this need to be public? Please try to limit the public surface

Reply:
Actually, it does! This code runs and keeps track of runtime, but the delay is computed with additional variables such as crash count, the amount of time a process runs to be considered healthy and the healthy crash amount. This is handled by the code that will use this package!

Reply:
Would it make sense to bring the [manageProcess function](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/453/files#diff-6e1a2f4e7ee6519eddd909e6b6cd806aR164) (and some of the related code) from your other PR into this package, or is there a reason why you've kept that in the receiver?

It does seem like a better boundary for this package would include everything related to managing and restarting processes.

Reply:
I think I need to keep lots of the logic in the parent code. 

The `prometheus_exec` receiver does things such as generate a new port (which is used by both the receiver and the subprocess with string templating and such),  and this port can/will change between every process crash. This means that the `Run()` function needs to be called at every crash with a different (modified) `SubprocessSettings` struct. 

The subprocess manager in its current form I think is still useful since it allows someone to start a process, and has a helper function to help the calling code compute exponential backoff. I see what you mean for the boundary but in my case I need the receiver to have more power on the subprocess starting since things often change between crashes. Also, I need to stop the associated receiver when the process crashes, for the same time that the process starts, so a more "black box" approach to the subprocess management isn't possible I reckon.

Reply:
Would it make more sense to add the configurable healthy fields to `SubprocessSettings` and have `GetDelay()` become a method to avoid requiring the user track them w/ each call?  It could also be more ideal if there were a notion of state/history that is managed and returned by `Run()` so that `GetDelay()` wouldn't require any args.  I'm not familiar w/ other subprocess managers/utilities in other languages having these notions so the interface is currently a little unintuitive to me, though I would like to consume this helper in contrib receivers.

Reply:
Yes, there is a need for a configurable subprocess manager. I won't be going forward with this PR (thanks for the comment, I'll be closing it), but [this issue](https://github.com/open-telemetry/opentelemetry-collector/issues/1492) exists to track the requirements needed for a subprocess manager that suits everyone's needs.